### PR TITLE
Control flow for element access expressions

### LINKED ISSUE #28081: Type guard using square bracket notation does not narrow type
**TypeScript Version:**  3.2.0-dev.20181023

**Search Terms:**
type guard square bracket notation narrowing

**Code**

```ts
interface ConfigType {
    [key: string]: boolean | { prop: string };
}

const config: ConfigType = {
    works: { prop: 'foo' },
    'works-not': { prop: 'bar' },
    bool: true,
};

if (typeof config.works !== 'boolean') {
    config.works.prop = 'test';
}

if (typeof config['works-not'] !== 'boolean') {
    config['works-not'].prop = 'test'; // [ts] Property 'prop' does not exist on type 'false'.
}
```

**Expected behavior:**
The type guard using square bracket notation narrows the type in the same way the type guard using dot notation does, resulting in no type errors in the above example.

**Actual behavior:**
The type guard specifically using square bracket notation does not narrow the type.

**Playground Link:** 
[Link](http://www.typescriptlang.org/play/#src=interface%20ConfigType%20%7B%0D%0A%20%20%20%20%5Bkey%3A%20string%5D%3A%20boolean%20%7C%20%7B%20prop%3A%20string%20%7D%3B%0D%0A%7D%0D%0A%0D%0Aconst%20config%3A%20ConfigType%20%3D%20%7B%0D%0A%20%20%20%20works%3A%20%7B%20prop%3A%20'foo'%20%7D%2C%0D%0A%20%20%20%20'works-not'%3A%20%7B%20prop%3A%20'bar'%20%7D%2C%0D%0A%20%20%20%20bool%3A%20true%2C%0D%0A%7D%3B%0D%0A%0D%0Aif%20(typeof%20config.works%20!%3D%3D%20'boolean')%20%7B%0D%0A%20%20%20%20config.works.prop%20%3D%20'test'%3B%0D%0A%7D%0D%0A%0D%0Aif%20(typeof%20config%5B'works-not'%5D%20!%3D%3D%20'boolean')%20%7B%0D%0A%20%20%20%20config%5B'works-not'%5D.prop%20%3D%20'test'%3B%20%2F%2F%20%5Bts%5D%20Property%20'prop'%20does%20not%20exist%20on%20type%20'false'.%0D%0A%7D%0D%0A)

**Related Issues:** 
https://github.com/Microsoft/TypeScript/issues/10530

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #28081

I did some additional performance investigation, using the same synthetic stress test as above with 3 variants:

## Variant 1 - Literal, Narrowing

```ts 
function stress(o: { [s: string]: "aa" | "ab" | "ac" | ... }) {
  o["x"] = o["z"] === "de" ? "ff" : "ip";
  // many more lines like this...
  return [o['x'], o['y'], o['z']] as const
}
```

## Variant 2 - Primitive, Non-Narrowing

```ts
function stress(o: { [s: string]: "aa" | "ab" | "ac" | ... }, x: string, y: string, z: string) {
  o[x] = o[z] === "de" ? "ff" : "ip";
  // many more lines like this...
  return [o[x], o[y], o[z]] as const
}
```

## Variant 3 - Literal Type, Non-Narrowing

```ts
function stress(o: { [s: string]: "aa" | "ab" | "ac" | ... }, x: 'x', y: 'y', z: 'z') {
  o[x] = o[z] === "de" ? "ff" : "ip";
  // many more lines like this...
  return [o[x], o[y], o[z]] as const
}
```

The most important result is that in real-world scenarios, this change never makes a significant difference in compile time: the second two variants don't actually create flow nodes in the binder, so control flow quits early as there's nothing to analyze. This means that no narrowing happens, of course. Inserting flow nodes only for element accesses with literal access expressions was a performance decision from quite some time ago. As far as I know, we have no reason to revisit that decision.

The first variant does narrow, but only 1000 lines. After that, control flow throws an error "The containing function or module body is too large for control flow analysis."

So I cut the stress tests down to 1000 lines and did a before/after comparison of them:
Selection_046

As you can see, the worst case is less than 50% increase per very-long function. I think that's acceptable given the improvement in narrowing for the common case.

### CODE REVIEW

**Comment by sandersn** [comment]
@typescript-bot run perf

**Comment by sandersn** [comment]
@typescript-bot perf test this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -20503,7 +20503,45 @@ namespace ts {
...
+                return checkIndexedAccessIndexType(indexedAccessType, node);
+            }
+            // If strict null checks and strict property initialization checks are enabled, if we have
+            // a this.xxx property access, if the property is an instance property without an initializer,
+            // and if we are in a constructor of the same class as the property declaration, assume that
+            // the property is uninitialized at the top of the control flow.
+            let assumeUninitialized = false;
+            if (strictNullChecks && strictPropertyInitialization && node.expression.kind === SyntaxKind.ThisKeyword) {
+                const declaration = prop && prop.valueDeclaration;
+                if (declaration && isInstancePropertyWithoutInitializer(declaration)) {
```
Comment (weswigham) [inline src/compiler/checker.ts]: Could this be unified with the similar code in `checkPropertyAccessExpressionOrQualifiedName`? Maybe a shared worker, like `getFlowTypeOfReferenceWithImpliedInitialType`?

Comment (sandersn) [reply src/compiler/checker.ts]: yes! It is basically a cut+paste+rename, so it is next on my list after I find out how much this hurts performance.

**Comment by sandersn** [comment]
Hm, no change in perf? That can't be right. I'm going to add a micro-benchmark to the perf test suite and report its results.

**Comment by sandersn** [comment]
I just wrote a stress test for this code. It's of the form:

```ts
function stress(o: { [s: string]: "aa" | "ab" | "ac" | ... }) {
  o["x"] = o["z"] === "de" ? "ff" : "ip";
  // many more lines like this...
  return [o['x'], o['y'], o['z']] as const
}
```

It scales quite well; check time at 100 lines was 0.28 seconds, at 1000 lines it was 0.5 and at 10,000 it was 0.6. Memory increases from 128 MB to 157 MB which is not bad.

What didn't work as well was code like this:

```ts
function stress(o: { [s: string]: "aa" | "ab" | "ac" | ... }, x: string, y: string, z: string) {
  o[x] = o[z] === "de" ? "ff" : "ip";
  // many more lines like this...
  return [o[x], o[y], o[z]] as const
}
```

It scales linearly or a little worse, up to 5.5 seconds and 178 MB for 16,000 lines.

I'll see if I can get a newer code base or two added to our performance suite although I'm not sure what is a good test for this change.

**Comment by sandersn** [comment]
Update: with more data points, it's pretty clear that the second example scales exponentially, not linearly.

**Comment by sandersn** [comment]
Well, this PR is as ready to go as it'll ever be. Any opinions @ahejlsberg?

@rbuckton any idea for good code bases that would show off real-world perf? My synthetic benchmark is intentionally unrepresentative.

**Comment by sandersn** [comment]
From the brief discussion in the 6/28 design meeting (#32357), @rbuckton raised the question of performance on a stress test, which is a variant of the first where identifiers of literal type are substituted for literals. I expect performance to be about the same.


```ts
function stress(o: { [s: string]: "aa" | "ab" | "ac" | ... }, x: 'x', y: 'y', z: 'z') {
  o[x] = o[z] === "de" ? "ff" : "ip";
  // many more lines like this...
  return [o[x], o[y], o[z]] as const
}
```

**Comment by sandersn** [comment]
I did some additional performance investigation, using the same synthetic stress test as above with 3 variants:

## Variant 1 - Literal, Narrowing

```ts 
function stress(o: { [s: string]: "aa" | "ab" | "ac" | ... }) {
  o["x"] = o["z"] === "de" ? "ff" : "ip";
  // many more lines like this...
  return [o['x'], o['y'], o['z']] as const
}
```

## Variant 2 - Primitive, Non-Narrowing

```ts
function stress(o: { [s: string]: "aa" | "ab" | "ac" | ... }, x: string, y: string, z: string) {
  o[x] = o[z] === "de" ? "ff" : "ip";
  // many more lines like this...
  return [o[x], o[y], o[z]] as const
}
```

## Variant 3 - Literal Type, Non-Narrowing

```ts
function stress(o: { [s: string]: "aa" | "ab" | "ac" | ... }, x: 'x', y: 'y', z: 'z') {
  o[x] = o[z] === "de" ? "ff" : "ip";
  // many more lines like this...
  return [o[x], o[y], o[z]] as const
}
```

The most important result is that in real-world scenarios, this change never makes a significant difference in compile time: the second two variants don't actually create flow nodes in the binder, so control flow quits early as there's nothing to analyze. This means that no narrowing happens, of course. Inserting flow nodes only for element accesses with literal access expressions was a performance decision from quite some time ago. As far as I know, we have no reason to revisit that decision.

The first variant does narrow, but only 1000 lines. After that, control flow throws an error "The containing function or module body is too large for control flow analysis."

So I cut the stress tests down to 1000 lines and did a before/after comparison of them:
Selection_046

As you can see, the worst case is less than 50% increase per very-long function. I think that's acceptable given the improvement in narrowing for the common case.

**Pull request merged**